solidity_contracts/src/CairoPrecompiles/DualVmToken.sol

```diff
@@ -129,44 +220,115 @@ contract DualVmToken {
+        return true;
+    }
+
+    function _transfer(uint256 to, uint256 amount) private {
+        if (to >= STARKNET_FIELD_PRIME) {
+            revert InvalidStarknetAddress();
+        }
         // Split amount in [low, high]
         uint128 amountLow = uint128(amount);
         uint128 amountHigh = uint128(amount >> 128);
 
         uint256[] memory transferCallData = new uint256[](3);
-        transferCallData[0] = toStarknetAddress;
+        transferCallData[0] = to;
         transferCallData[1] = uint256(amountLow);
         transferCallData[2] = uint256(amountHigh);
 
         starknetToken.delegatecallCairo("transfer", transferCallData);
-
-        emit Transfer(msg.sender, to, amount);
-        return true;
     }
 
-    function transferFrom(address from, address to, uint256 amount) public virtual returns (bool) {
+    function transferFrom(address from, address to, uint256 amount) external returns (bool) {
+        uint256[] memory toAddressCalldata = new uint256[](1);
+        toAddressCalldata[0] = uint256(uint160(to));
+        uint256 toStarknetAddress =
+            abi.decode(kakarot.staticcallCairo("get_starknet_address", toAddressCalldata), (uint256));
+
         uint256[] memory fromAddressCalldata = new uint256[](1);
         fromAddressCalldata[0] = uint256(uint160(from));
         uint256 fromStarknetAddress =
             abi.decode(kakarot.staticcallCairo("get_starknet_address", fromAddressCalldata), (uint256));
 
+        _transferFrom(fromStarknetAddress, toStarknetAddress, amount);
+
+        emit Transfer(from, to, amount);
+        return true;
+    }
+
+    /// @dev Transfer tokens from a starknet address
+    /// @param from The starknet address to transfer the tokens from
+    /// @param to The evm address to transfer the tokens to
+    /// @param amount The amount of tokens to transfer
+    /// @return True if the transfer was successful
+    function transferFromStarknetAddressFrom(uint256 from, address to, uint256 amount) external returns (bool) {
         uint256[] memory toAddressCalldata = new uint256[](1);
-        toAddressCalldata[0] = uint256(uint160(to));
+        uint256 toAddressUint256 = uint256(uint160(to));
+        toAddressCalldata[0] = toAddressUint256;
         uint256 toStarknetAddress =
             abi.decode(kakarot.staticcallCairo("get_starknet_address", toAddressCalldata), (uint256));
 
+        _transferFrom(from, toStarknetAddress, amount);
+
+        emit TransferStarknet(from, toAddressUint256, amount);
+        return true;
+    }
+
+    /// @dev Transfer from tokens to a starknet address
+    /// @param from The evm address to transfer the tokens from
+    /// @param to The starknet address to transfer the tokens to
+    /// @param amount The amount of tokens to transfer
+    /// @return True if the transfer was successful
+    function transferFromStarknetAddressTo(address from, uint256 to, uint256 amount) external returns (bool) {
```

Comment:
@ClementWalter those functions are using starknetAddress as input 
We want to add this explicitly in the function name to avoid mistakes from users.
But we are not satisfied with this naming like `transferFromStarknetAddressFromAndTo`
Do you have a suggestion ?

Reply:
can't we just use the type to distinguish? `address` for EVM and `uint256` for starknet? and keep only one signature

Reply:
Not sure how overloading will be handled with python. We are modifying the `functions` attribute of contracts to wrap them with the `wrap_kakarot` but they will share the same name (type is not attached to the attribute `functions`). Web3 do not recognize different overloaded functions with this method.
Though anyway it was better to have explicit mention of the fact the address are particular and ensure devs do not make mistakes.

Reply:
Maybe with selectors we can replace `abi = self.get_function_by_name(fun).abi` by getting selectors instead?

Reply:
Will try that and combine with User-defined Value Types from solidity 
https://docs.soliditylang.org/en/v0.8.27/types.html#user-defined-value-types

Reply:
https://web3py.readthedocs.io/en/v6.10.0/web3.contract.html#invoke-ambiguous-contract-functions-example
It seems the only way to deal with overloading is to use `get_function_by_signature`.
hence using <contracts>.functions is not possible as this method is returning the name only. Will think of a way to bypass this limitation

Reply:
one solution would be: 
* make a new `_wrap_kakarot_by_signature` which would mimic the `_wrap_kakarot` but would call `get_function_by_signature` instead of `get_function_by_name` to get the abi and other informations
* adapt the function to do  `await _wrap_kakarot_by_signature(<contract>, <func_signature>, <args>)`

Trying to implement that but no success for now